File: clients/kotlin/src/main/kotlin/dev/zmr/ZmrClient.kt
package dev.zmr

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.Closeable
import java.io.InputStreamReader
import java.io.OutputStreamWriter

class ZmrRpcException(
    val code: Int,
    message: String,
    val publicCode: String? = null
) : RuntimeException(message)

data class TraceDiscoverOptions(
    val includeActions: Boolean = false,
    val validate: Boolean = false,
    val force: Boolean = false,
    val name: String? = null,
    val appId: String? = null
)

class ZmrClient(
    private val command: List<String> = listOf("zmr", "serve", "--transport", "stdio")
) : Closeable {
    private var nextId = 1
    private val process = ProcessBuilder(command).redirectError(ProcessBuilder.Redirect.INHERIT).start()
    private val input = BufferedWriter(OutputStreamWriter(process.outputStream))
    private val output = BufferedReader(InputStreamReader(process.inputStream))

    fun createSession(): String = call("session.create")

    fun snapshot(): String = call("observe.snapshot")

    fun semanticSnapshot(): String = call("observe.semanticSnapshot")

    fun assertHealthy(timeoutMs: Long? = null): String {
        val params = timeoutMs?.let { "{\"timeoutMs\":$it}" } ?: "{}"
        return call("assert.healthy", params)
    }

    fun validateScenario(path: String): String =
        call("scenario.validate", """{"path":"${escapeJson(path)}"}""")

    fun explainTrace(): String = call("trace.explain")

    fun discoverTrace(out: String, options: TraceDiscoverOptions = TraceDiscoverOptions()): String {
        val fields = mutableListOf(""""out":"${escapeJson(out)}"""")
        if (options.includeActions) fields.add(""""includeActions":true""")
        if (options.validate) fields.add(""""validate":true""")
        if (options.force) fields.add(""""force":true""")
        options.name?.let { fields.add(""""name":"${escapeJson(it)}"""") }
        options.appId?.let { fields.add(""""appId":"${escapeJson(it)}"""") }
        return call("trace.discover", "{${fields.joinToString(",")}}")
    }

    @Synchronized
    fun call(method: String, paramsJson: String? = null): String {
        val id = nextId++
        val params = paramsJson?.let { "," + "\"params\":" + it } ?: ""
        input.write("{\"jsonrpc\":\"2.0\",\"id\":$id,\"method\":\"$method\"$params}")
        input.newLine()
        input.flush()
        val response = output.readLine() ?: error("zmr closed stdout")
        if (hasTopLevelKey(response, "error")) {
            throw ZmrRpcException(
                code = extractNumber(response, "code") ?: -32000,
                message = extractString(response, "message").ifEmpty { "ZMR JSON-RPC error" },
                publicCode = extractString(response, "publicCode").ifEmpty { null }
            )
        }
        return response
    }

    override fun close() {
        runCatching { call("session.close") }
        runCatching { input.close() }
        process.destroy()
    }
}

private fun extractString(json: String, key: String): String {
    val pattern = """"$key"\s*:\s*"([^"]*)"""".toRegex()
    return pattern.find(json)?.groupValues?.get(1) ?: ""
}

private fun extractNumber(json: String, key: String): Int? {
    val pattern = """"$key"\s*:\s*(-?[0-9]+)""".toRegex()
    return pattern.find(json)?.groupValues?.get(1)?.toIntOrNull()
}

private fun hasTopLevelKey(json: String, key: String): Boolean {
    var depth = 0
    var inString = false
    var escaped = false
    var stringStart = 0
    var i = 0
    while (i < json.length) {
        val ch = json[i]
        if (inString) {
            when {
                escaped -> escaped = false
                ch == '\\' -> escaped = true
                ch == '"' -> {
                    inString = false
                    if (depth == 1 && json.substring(stringStart, i) == key) {
                        var j = i + 1
                        while (j < json.length && json[j].isWhitespace()) j += 1
                        if (j < json.length && json[j] == ':') return true
                    }
                }
            }
        } else {
            when (ch) {
                '"' -> {
                    inString = true
                    stringStart = i + 1
                }
                '{', '[' -> depth += 1
                '}', ']' -> depth -= 1
            }
        }
        i += 1
    }
    return false
}

private fun escapeJson(value: String): String =
    value.replace("\\", "\\\\").replace("\"", "\\\"")

File: clients/swift/Sources/ZMRClient/ZMRClient.swift
import Foundation

public enum ZMRError: Error {
    case processNotStarted
    case invalidResponse
    case rpcError([String: Any])
}

public struct TraceDiscoverOptions {
    public var includeActions: Bool
    public var validate: Bool
    public var force: Bool
    public var name: String?
    public var appId: String?

    public init(
        includeActions: Bool = false,
        validate: Bool = false,
        force: Bool = false,
        name: String? = nil,
        appId: String? = nil
    ) {
        self.includeActions = includeActions
        self.validate = validate
        self.force = force
        self.name = name
        self.appId = appId
    }
}

public final class ZMRClient {
    private let process: Process
    private let input: FileHandle
    private let output: FileHandle
    private var nextID = 1

    public init(executable: String = "zmr", arguments: [String] = ["serve", "--transport", "stdio"]) {
        let process = Process()
        if executable.contains("/") {
            process.executableURL = URL(fileURLWithPath: executable)
            process.arguments = arguments
        } else {
            process.executableURL = URL(fileURLWithPath: "/usr/bin/env")
            process.arguments = [executable] + arguments
        }

        let stdinPipe = Pipe()
        let stdoutPipe = Pipe()
        process.standardInput = stdinPipe
        process.standardOutput = stdoutPipe
        process.standardError = FileHandle.standardError

        self.process = process
        self.input = stdinPipe.fileHandleForWriting
        self.output = stdoutPipe.fileHandleForReading
    }

    public func start() throws {
        try process.run()
    }

    public func close() {
        _ = try? call("session.close")
        input.closeFile()
        if process.isRunning {
            process.terminate()
        }
    }

    @discardableResult
    public func call(_ method: String, params: [String: Any]? = nil) throws -> Any {
        guard process.isRunning else { throw ZMRError.processNotStarted }
        let id = nextID
        nextID += 1

        var request: [String: Any] = ["jsonrpc": "2.0", "id": id, "method": method]
        if let params {
            request["params"] = params
        }
        let data = try JSONSerialization.data(withJSONObject: request, options: [])
        input.write(data)
        input.write(Data([0x0a]))

        let line = try readLineData()
        let object = try JSONSerialization.jsonObject(with: line, options: [])
        guard let response = object as? [String: Any] else { throw ZMRError.invalidResponse }
        if let error = response["error"] as? [String: Any] {
            throw ZMRError.rpcError(error)
        }
        guard let result = response["result"] else { throw ZMRError.invalidResponse }
        return result
    }

    public func createSession() throws {
        _ = try call("session.create")
    }

    public func snapshot() throws -> [String: Any] {
        guard let result = try call("observe.snapshot") as? [String: Any] else {
            throw ZMRError.invalidResponse
        }
        return result
    }

    public func semanticSnapshot() throws -> [String: Any] {
        guard let result = try call("observe.semanticSnapshot") as? [String: Any] else {
            throw ZMRError.invalidResponse
        }
        return result
    }

    public func assertHealthy(timeoutMs: Int? = nil) throws -> Bool {
        var params: [String: Any] = [:]
        if let timeoutMs {
            params["timeoutMs"] = timeoutMs
        }
        guard let result = try call("assert.healthy", params: params) as? Bool else {
            throw ZMRError.invalidResponse
        }
        return result
    }

    public func validateScenario(path: String) throws -> [String: Any] {
        guard let result = try call("scenario.validate", params: ["path": path]) as? [String: Any] else {
            throw ZMRError.invalidResponse
        }
        return result
    }

    public func discoverTrace(out: String, options: TraceDiscoverOptions = TraceDiscoverOptions()) throws -> [String: Any] {
        var params: [String: Any] = ["out": out]
        if options.includeActions {
            params["includeActions"] = true
        }
        if options.validate {
            params["validate"] = true
        }
        if options.force {
            params["force"] = true
        }
        if let name = options.name {
            params["name"] = name
        }
        if let appId = options.appId {
            params["appId"] = appId
        }
        guard let result = try call("trace.discover", params: params) as? [String: Any] else {
            throw ZMRError.invalidResponse
        }
        return result
    }

    public func explainTrace() throws -> [String: Any] {
        guard let result = try call("trace.explain", params: [:]) as? [String: Any] else {
            throw ZMRError.invalidResponse
        }
        return result
    }

    private func readLineData() throws -> Data {
        var data = Data()
        while true {
            let byte = output.readData(ofLength: 1)
            if byte.isEmpty {
                throw ZMRError.invalidResponse
            }
            if byte[0] == 0x0a {
                return data
            }
            data.append(byte)
        }
    }
}
